frontends/PyCDE/src/pycde/types.py

```diff
@@ -42,17 +46,32 @@ def maybe_create_alias(self, inner_type, name):
       if name in self.registered_aliases:
         if alias != self.registered_aliases[name]:
           raise RuntimeError(
-              f"Re-defining type alias for {name}! "\
-              f"Given: {inner_type}, "\
-              f"existing: {self.registered_aliases[name].inner_type}"
-          )
+              f"Re-defining type alias for {name}! "
+              f"Given: {inner_type}, "
+              f"existing: {self.registered_aliases[name].inner_type}")
         return self.registered_aliases[name]
 
       self.registered_aliases[name] = alias
       return alias
 
     return inner_type
 
+  def declare_types(self, mod):
+    if not self.registered_aliases:
+      return
+
+    if not self.type_scopes.get(mod):
+      with mlir.ir.InsertionPoint.at_block_begin(mod.body):
+        self.type_scopes[mod] = hw.TypeScopeOp.create(self.TYPE_SCOPE)
+        self.declared_aliases[mod] = set([])
+
+    with mlir.ir.InsertionPoint(self.type_scopes[mod].body):
+      for (name, type) in self.registered_aliases.items():
```

Comment:
Is it possible to get this information out of the IR?

I'm very sensitive to duplicate state since it is very easy for for them to get out of sync.